sm/enclave.c

```diff
@@ -717,3 +732,176 @@ enclave_ret_code get_sealing_key(uintptr_t sealing_key, uintptr_t key_ident,
+
+     	if(hdr->send_uid == uid){
+        	//Check if the message is bigger than the buffer. 
+        	if(hdr->size > msg_size){
+            		spinlock_unlock(&(mailbox->lock));
+            		return 1; 
+		}
+
+        	memcpy(buf, hdr->data, msg_size); 
+
+        	//Clear the message from the mailbox
+        	memset(hdr->data, 0, hdr->size);
+        	memset(hdr, 0, sizeof(struct mailbox_header));
+        	memcpy(hdr, ptr + hdr_size + sizeof(struct mailbox_header), mailbox->size - (size + sizeof(struct mailbox_header) + hdr_size)); 
+
+        	mailbox->size -= hdr_size + sizeof(struct mailbox_header); 
+        	spinlock_unlock(&(mailbox->lock));
+		return 0; 
+     	}
+  		size += sizeof(struct mailbox_header) + hdr_size;
+    		ptr += sizeof(struct mailbox_header) + hdr_size;    
+    		hdr = (struct mailbox_header *) ptr;
+	}
+	//Release lock on mailbox 
+  	spinlock_unlock(&(mailbox->lock));
+	return 1;
+
+}
+
+
+enclave_ret_code send_msg(enclave_id eid, size_t uid, void *buf, size_t msg_size){
+   struct mailbox *mbox = (void *) 0; 
+
+   for(int eid=0; eid<ENCL_MAX; eid++)
+  {
+    if(ENCLAVE_EXISTS(eid) && enclaves[eid].uid == uid){
+      mbox = &enclaves[eid].mailbox;
+      break; 
+    }
+  }
+
+   //Check if the mailbox is registered
+   if(!mbox){
+      return 1;  
+   }
+
+   spinlock_lock(&(mbox->lock));
+ 
+   //Check if the message + header can fit in the mailbox. 
+   if(mbox->capacity - mbox->size < msg_size + sizeof(struct mailbox_header)){
+      spinlock_unlock(&(mbox->lock));
+      return 0; 
+   }
+
+   struct mailbox_header hdr;
+   hdr.send_uid = enclaves[eid].uid; 
+   hdr.size = msg_size;
+ 
+   memcpy(mbox->data + mbox->size, &hdr, sizeof(hdr));
+   memcpy(mbox->data + mbox->size + sizeof(hdr), buf, msg_size);  
+   mbox->size += msg_size + sizeof(hdr);  
+
+   spinlock_unlock(&(mbox->lock));
+   
+   return 0;
+}
+
+enclave_ret_code mem_share(enclave_id eid, size_t uid, uintptr_t *enclave_addr, uintptr_t *enclave_size){
+   struct enclave *grantee = (void *) 0; 
+   int grantee_eid; 
+   //return ENCLAVE_SUCCESS; 
+   //Find the enclave with the corresponding uid
+   for(int eid=0; eid<ENCL_MAX; eid++)
+  {
+    if(ENCLAVE_EXISTS(eid) && enclaves[eid].uid == uid){
+      grantee = &enclaves[eid];
+      grantee_eid = eid; 
+      break;
+    }
+  }   	
+
+  if(!grantee){
+      return 1; 
+  }
+
+  //Set PMP of the granter enclave with grantee
+  int memid;
+  for(memid=0; memid < ENCLAVE_REGIONS_MAX; memid++) {
+    if(enclaves[eid].regions[memid].type == REGION_EPM) {
+      /* Find empty memory region slot in grantee */
+      for(int grantee_memid = 0; grantee_memid < ENCLAVE_REGIONS_MAX; grantee_memid++){
+	if(grantee->regions[grantee_memid].type == REGION_INVALID){
+		grantee->regions[grantee_memid].pmp_rid = enclaves[eid].regions[memid].pmp_rid;
+		grantee->regions[grantee_memid].type = enclaves[eid].regions[memid].type;
+		break;
+	}
+      }
+    }
+  }
+
+  if(enclave_addr)
```

Comment:
can you use mprv instead of passing physical address to this SBI?
with mprv, you can pass in the runtime's virtual address and the SM will use satp to translate the VA.